File: ComExamples/Com232MeshTest/Com232MeshTest.cpp
// Copyright © 2026 Khrustal & Mann
//              MELBOURNE, VICTORIA, AUSTRALIA, 3000
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
//
// Com232MeshTest.cpp  (COM -- TargetCom)
//
// SINGLE-PROCESS, TWO-HUB RS-232 serial connection probe, over COM.
//
// PREREQUISITE (unchanged from both other trees): a com0com virtual null-modem
// pair COM5<->COM6
//     setupc install PortName=COM5 PortName=COM6
// Without it this reports SETUP and exits 1.
//
// The port number used to cross as a LONG so that VT_I4 could carry it and the
// COM layer could range-check it before the facade saw a truncated `short`.
// There is no numeric parameter left to protect: the port is text inside
// "serial://COM5", and the check moved to the one place the grammar is parsed.
// A bad port is now P2PF_E_ENDPOINT out of the facade rather than E_INVALIDARG
// out of the COM layer -- the same mistake caught in the same call, by a rule
// that every one of the three trees now shares instead of just this one.
//
// Verdict by EXIT CODE:
//   0 = SUCCESS : the server hub received the client's message over the wire.
//   3 = TIMEOUT : handshake or delivery did not complete in time.
//   1 = SETUP   : no com0com pair, or the COM server is not registered.

#include "ComHarness.h"

static const unsigned short kServerComPort = 5;
static const unsigned short kClientComPort = 6;
static const wchar_t* const kServerAddr    = L"Com232Mesh.Server";
static const wchar_t* const kClientAddr    = L"Com232Mesh.Client";
static const wchar_t* const kTopic         = L"mesh";

int main ( )
{
    com::InitConsole();
    wprintf ( L"=== Com232MeshTest (COM) - two hubs, one process, RS-232 ===\n" );
    wprintf ( L"Server : COM%d   Client : COM%d   (com0com null-modem pair required)\n\n",
              (int)kServerComPort, (int)kClientComPort );
    fflush ( stdout );

    com::Apartment sta;
    if ( !sta.ok() ) return com::EXIT_SETUP;

    com::Network net;
    if ( !net.ok() ) return com::SetupFailure ( L"CoCreateInstance(TargetCom.P2PNetwork)", net.hr() );

    com::Gate gDone;
    com::Hub  server, client;

    HRESULT hr = net.createHub ( kServerAddr, server );
    if ( FAILED(hr) ) return com::SetupFailure ( L"CreateHub(server)", hr );

    server.onTopic ( kTopic, [&] ( const com::Message& m )
    {
        com::LogMessage ( L"SERVER", L"message", m.source, m.text() );
        gDone.open();
    });
    server.onPeerUp   ( [] ( LPCWSTR peer ) { com::Log ( L"SERVER", L"peer up   : %s", peer ); } );
    server.onPeerDown ( [] ( LPCWSTR peer ) { com::Log ( L"SERVER", L"peer down : %s", peer ); } );
    server.onError    ( [] ( LPCWSTR what ) { com::Log ( L"SERVER", L"error     : %s", what ); } );

    // Armed first: a serial:// dial does not retry.
    hr = server.listen ( kClientAddr, com::Serial ( kServerComPort ).c_str() );
    if ( FAILED(hr) )
    {
        com::Log ( L"SERVER", L"SETUP: Listen(serial://COM%d) -> %s",
                   (int)kServerComPort, com::HrName ( hr ) );
        wprintf ( L"\nNo com0com pair on COM%d/COM%d? Install one with:\n"
                  L"    setupc install PortName=COM%d PortName=COM%d\n",
                  (int)kServerComPort, (int)kClientComPort,
                  (int)kServerComPort, (int)kClientComPort );
        fflush ( stdout );
        return com::EXIT_SETUP;
    }
    com::Log ( L"SERVER", L"COM%d armed for '%s'", (int)kServerComPort, kClientAddr );

    hr = net.createHub ( kClientAddr, client );
    if ( FAILED(hr) ) return com::SetupFailure ( L"CreateHub(client)", hr );

    client.onPeerUp ( [&] ( LPCWSTR peer )
    {
        com::Log ( L"CLIENT", L"peer up   : %s - serial link ready", peer );
        HRESULT hrSend = client.sendText ( kServerAddr, kTopic,
                                           L"Hello over RS-232, through COM!" );
        com::Log ( L"CLIENT", L"SendText  : %s", com::HrName ( hrSend ) );
    });
    client.onPeerDown ( [] ( LPCWSTR peer ) { com::Log ( L"CLIENT", L"peer down : %s", peer ); } );
    client.onError    ( [] ( LPCWSTR what ) { com::Log ( L"CLIENT", L"error     : %s", what ); } );

    hr = client.connect ( kServerAddr, com::Serial ( kClientComPort ).c_str() );
    if ( FAILED(hr) )
    {
        com::Log ( L"CLIENT", L"SETUP: Connect(serial://COM%d) -> %s",
                   (int)kClientComPort, com::HrName ( hr ) );
        return com::EXIT_SETUP;
    }
    com::Log ( L"CLIENT", L"dialled from COM%d", (int)kClientComPort );

    // The range check, now in the endpoint parser rather than the COM layer.
    // "serial://COM0" is a legal-looking string with an illegal port in it, and
    // the parser insists the whole segment be numeric AND in range -- so this is
    // rejected for the same reason "serial://COMx" would be.
    HRESULT hrBad = server.listen ( L"Com232Mesh.Nope", com::Serial ( 0 ).c_str() );
    com::Log ( L"MAIN", L"Listen(serial://COM0) -> %s (range-checked in the parser)",
               com::HrName ( hrBad ) );

    com::Log ( L"MAIN", L"waiting up to 15s for login + delivery (pumping)..." );
    bool ok = gDone.wait ( 15000 ) && ( hrBad == com::E_ENDPOINT );

    com::Log ( L"MAIN", L"shutdown begin" );
    return com::Verdict ( ok,
                          L"server received the client's message over RS-232",
                          L"no message delivered (handshake did not complete)" );
}

File: dotNetExamples/Com232MeshTest/Com232MeshTest.cs
// Copyright © 2026 Khrustal & Mann
//              MELBOURNE, VICTORIA, AUSTRALIA, 3000
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
//
// Com232MeshTest.cs  (C# -- TargetCom)
//
// SINGLE-PROCESS, TWO-HUB RS-232 serial connection probe, over COM from managed
// code.
//
// PREREQUISITE (unchanged from all three other trees): a com0com virtual
// null-modem pair COM5<->COM6
//     setupc install PortName=COM5 PortName=COM6
// Without it this reports SETUP and exits 1.
//
// The port used to cross as a LONG -- a plain `int` here -- so the COM layer
// could range-check it before the facade saw a truncated `short`. Since ABI 4
// there is no numeric parameter left: the port is text inside "serial://COM5"
// and the check moved into the endpoint parser, which every one of the four
// trees now shares instead of just the two COM ones.
//
// That move takes the *shape* of the error with it, and this harness is where
// that is visible. See the two-part section near the bottom.
//
// Verdict by EXIT CODE:
//   0 = SUCCESS : the server hub received the client's message over the wire.
//   3 = TIMEOUT : handshake or delivery did not complete in time.
//   1 = SETUP   : no com0com pair, or the COM server is not registered.

using System;
using System.Runtime.InteropServices;

using Com;
using TargetCom;

static class Com232MeshTest
{
    const int    ServerComPort = 5;
    const int    ClientComPort = 6;
    const string ServerAddr    = "Com232Mesh.Server";
    const string ClientAddr    = "Com232Mesh.Client";
    const string Topic         = "mesh";

    [STAThread]
    static int Main ()
    {
        Harness.InitConsole();
        Harness.Print("=== Com232MeshTest (C#) - two hubs, one process, RS-232 ===");
        Harness.Print("Server : COM{0}   Client : COM{1}   (com0com null-modem pair required)\n",
                      ServerComPort, ClientComPort);

        if (!Apartment.IsSta) return Harness.ExitSetup;

        int hr;
        using var net = Network.Create(out hr);
        if (net == null) return Harness.SetupFailure("CoCreateInstance(TargetCom.P2PNetwork)", hr);

        var gDone = new Gate();

        using var server = net.CreateHub(ServerAddr, out hr);
        if (server == null) return Harness.SetupFailure("CreateHub(server)", hr);

        server.OnTopic(Topic, m =>
        {
            Harness.LogMessage("SERVER", "message", m.Source, m.Text);
            gDone.Open();
        });
        server.OnPeerUp  (peer => Harness.Log("SERVER", "peer up   : {0}", peer));
        server.OnPeerDown(peer => Harness.Log("SERVER", "peer down : {0}", peer));
        server.OnError   (what => Harness.Log("SERVER", "error     : {0}", what));

        // Armed first: a serial:// dial does not retry.
        hr = server.Listen(ClientAddr, Endpoint.Serial(ServerComPort));
        if (Hr.Failed(hr))
        {
            Harness.Log("SERVER", "SETUP: Listen(serial://COM{0}) -> {1}", ServerComPort, Hr.Name(hr));
            Harness.Print("\nNo com0com pair on COM{0}/COM{1}? Install one with:\n"
                        + "    setupc install PortName=COM{0} PortName=COM{1}",
                          ServerComPort, ClientComPort);
            return Harness.ExitSetup;
        }
        Harness.Log("SERVER", "COM{0} armed for '{1}'", ServerComPort, ClientAddr);

        using var client = net.CreateHub(ClientAddr, out hr);
        if (client == null) return Harness.SetupFailure("CreateHub(client)", hr);

        client.OnPeerUp(peer =>
        {
            Harness.Log("CLIENT", "peer up   : {0} - serial link ready", peer);
            int hrSend = client.SendText(ServerAddr, Topic, "Hello over RS-232, through COM from C#!");
            Harness.Log("CLIENT", "SendText  : {0}", Hr.Name(hrSend));
        });
        client.OnPeerDown(peer => Harness.Log("CLIENT", "peer down : {0}", peer));
        client.OnError   (what => Harness.Log("CLIENT", "error     : {0}", what));

        hr = client.Connect(ServerAddr, Endpoint.Serial(ClientComPort));
        if (Hr.Failed(hr))
        {
            Harness.Log("CLIENT", "SETUP: Connect(serial://COM{0}) -> {1}", ClientComPort, Hr.Name(hr));
            return Harness.ExitSetup;
        }
        Harness.Log("CLIENT", "dialled from COM{0}", ClientComPort);

        // ---- PART 1: the range check, now in the endpoint parser -------------
        //
        // "serial://COM0" is a legal-looking string with an illegal port in it.
        // The parser insists the whole segment be numeric AND in range, so this
        // is rejected for the same reason "serial://COMx" would be.
        //
        // P2PF_E_ENDPOINT is a custom FACILITY_ITF code, so the CLR leaves it
        // alone and it arrives as an ordinary COMException -- which Com.Hub
        // turns back into an int. Contrast PART 2.
        int hrBad = server.Listen("Com232Mesh.Nope", Endpoint.Serial(0));
        Harness.Log("MAIN", "Listen(serial://COM0) -> {0} (range-checked in the parser)", Hr.Name(hrBad));

        // ---- PART 2: where the CLR still reshapes the error contract ---------
        //
        // This demonstration used to live on ListenSerial(port 0). The COM
        // layer's private Port() helper went away with its numeric parameters,
        // so its E_INVALIDARG went with it -- and CreateHub's empty-address
        // check is now the layer's ONLY remaining argument validation.
        //
        // The lesson is unchanged and still worth the lines: E_INVALIDARG is in
        // the CLR's HRESULT-to-exception mapping table, so the call site does
        // NOT get a COMException. It gets an ArgumentException, with the HRESULT
        // only recoverable via Marshal.GetHRForException. A C# client that
        // wrapped this API in `catch (COMException)` would sail past it and
        // crash. That is a property of the CLR, not of this interface, and no
        // amount of IDL tidying makes it go away.
        int hrEmpty = Hr.S_OK;
        try
        {
            net.Raw.CreateHub("");
            Harness.Log("MAIN", "CreateHub(\"\") was ACCEPTED - unexpected");
        }
        catch (COMException e)
        {
            hrEmpty = e.HResult;
            Harness.Log("MAIN", "CreateHub(\"\") -> COMException 0x{0:X8}", e.HResult);
        }
        catch (ArgumentException e)
        {
            hrEmpty = Marshal.GetHRForException(e);
            Harness.Log("MAIN", "CreateHub(\"\") -> {0} 0x{1:X8}  <- NOT a COMException; "
                              + "the CLR rewrote E_INVALIDARG", e.GetType().Name, hrEmpty);
        }

        Harness.Log("MAIN", "waiting up to 15s for login + delivery (pumping)...");
        bool ok = gDone.Wait(15000)
               && hrBad   == Hr.P2PF_E_ENDPOINT
               && hrEmpty == Hr.E_INVALIDARG;

        Harness.Log("MAIN", "shutdown begin");
        return Harness.Verdict(ok,
                               "server received the client's message over RS-232",
                               "no message delivered (handshake did not complete)");
    }
}
